### ISSUE TITLE
Reusing parser causes rules to show up that are not present

### ISSUE THREAD

**Issue by malwarefrank** [plyara/plyara#61]
I am using plyara to parse different rules and tracking statistics about them, but when I reuse the parser then the rules from the previous parse_string() call show up in the results of the subsequent unrelated parse_string() calls.

I can see this being useful in some cases, such as incrementally adding rules, but it is counter-intuitive to me that parse_string() is not stateless.  I suggest adding an option to make it stateless.

**Comment by rshipp** [comment]
Thanks for the issue and PR! This seems like a reasonable addition to me. Maintaining state across rules make plyara act more sensibly when dealing with files that have multiple rules, and your solution doesn't conflict with that.

I'll give some time for another maintainer to sign off or object, and otherwise I'll merge & ship this out in a few days.

**Comment by utkonos** [comment]
It appears that the full set of accumulators is not reset by the `clear()` methods in the PR. Before shipping this, we need to make sure that this resets everything correctly. Can you provide test cases that show the problem? This change will need an appropriate set of unit tests.

Is there a use case where this proposal isn't handled by instantiation of a new parser for each ruleset?

**Comment by utkonos** [comment]
Also, please lint the code with pydocstyle. This and the other linters we're using are not in the documentation, so I'm adding this to the README.

**Comment by utkonos** [comment]
One more item is the test coverage we have on the utils. I think that gap needs closing so that we know a change like this doesn't break those utils.

**Comment by malwarefrank** [comment]
Regarding your question, "Is there a use case where this proposal isn't handled by instantiation of a new parser for each ruleset?" I tried instantiating a new parser for each ruleset, but it took between 4 and 5 times as long to do so than when I tried just clearing the parser member variables.  That time difference adds up quickly for large-scale uses, such as realtime parsing of tens or hundreds of thousands of rulesets.

Regarding your other statements:
1. I will add test cases to the PR.
1. I'll try to better identify all of the accumulators that should be cleared and modify the PR accordingly.  Are there any specific ones that you know of that I missed?
1. I will run all the linters as they are listed in the .travis.yaml file and adjust the PR accordingly.
